language/testing-infra/functional-tests/src/testsuite.rs

```diff
@@ -65,31 +80,87 @@ fn print_stage(haystack: &str) -> bool {
         .unwrap_or(true)
 }
 
-// Runs all tests under the test/testsuite directory.
-pub fn functional_tests<TComp: Compiler>(
-    compiler: TComp,
-    path: &Path,
+fn write_horizontal_line(output: &mut Buffer, term_width: usize) -> std::io::Result<()> {
+    writeln!(
+        output,
+        "{}",
+        iter::repeat('=').take(term_width).collect::<String>()
+    )
+}
+
+fn write_test_header(
+    output: &mut Buffer,
+    term_width: usize,
+    test_file_path: &Path,
+) -> std::io::Result<()> {
+    writeln!(output)?;
+    write_horizontal_line(output, term_width)?;
+    writeln!(output, "{}", test_file_path.display())?;
+    writeln!(output)
+}
+
+fn check_or_update_expected_output(
+    bufwtr: &BufferWriter,
+    output: &mut Buffer,
+    term_width: usize,
+    test_file_path: &Path,
+    exp_file_path: &Path,
+    log: &EvaluationLog,
 ) -> datatest_stable::Result<()> {
-    let input = read_to_string(path)?;
+    // REVIEW: "\n" is probably not very portable.
+    let text = log
+        .to_text_for_matching()
+        .into_iter()
+        .enumerate()
+        .map(|(idx, entry)| format!("[{}] {}", idx, entry))
+        .join("\n");
```

Comment:
You want an extra \n here at the end. I imagine eventually some linter is going to yell for not having a newline at the end of the file